File: square-commons/src/main/java/com/raxdenstudios/square/interceptor/commons/toolbar/ToolbarActivityInterceptor.java
package com.raxdenstudios.square.interceptor.commons.toolbar;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.raxdenstudios.square.interceptor.ActivityInterceptor;

/**
 * Created by agomez on 21/05/2015.
 */
public class ToolbarActivityInterceptor extends ActivityInterceptor<ToolbarInterceptorCallback> implements ToolbarInterceptor {

    public ToolbarActivityInterceptor(@NonNull AppCompatActivity activity) {
        super(activity);
    }

    public ToolbarActivityInterceptor(@NonNull AppCompatActivity activity, @NonNull ToolbarInterceptorCallback callback) {
        super(activity, callback);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Toolbar toolbar = getCallback().onCreateToolbarView(savedInstanceState);
        if (toolbar != null) {
            if (getActivity() instanceof AppCompatActivity) {
                AppCompatActivity compatActivity = ((AppCompatActivity) getActivity());
                compatActivity.setSupportActionBar(toolbar);
                ActionBar actionBar = compatActivity.getSupportActionBar();
                if (actionBar != null) {
                    actionBar.setDisplayShowTitleEnabled(false);
                }
            }
            toolbar.setOnMenuItemClickListener(new Toolbar.OnMenuItemClickListener() {
                @Override
                public boolean onMenuItemClick(MenuItem item) {
                    return getActivity().onOptionsItemSelected(item);
                }
            });
            getCallback().onToolbarViewCreated(toolbar);
        }
    }

}

File: square-commons/src/main/java/com/raxdenstudios/square/interceptor/commons/toolbar/ToolbarActivityInterceptor.kt
package com.raxdenstudios.square.interceptor.commons.toolbar

import android.app.Activity
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar

import com.raxdenstudios.square.interceptor.ActivityInterceptor

/**
 * Created by agomez on 21/05/2015.
 *
    <android.support.constraint.ConstraintLayout
        xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:app="http://schemas.android.com/apk/res-auto"
        xmlns:tools="http://schemas.android.com/tools"
        android:id="@+id/activity_main"
        android:layout_width="match_parent"
        android:layout_height="match_parent">

        <android.support.v7.widget.Toolbar
            android:id="@+id/toolbar_view"
            android:layout_width="match_parent"
            android:layout_height="?attr/actionBarSize"
            android:background="@color/colorPrimary"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toTopOf="parent">

            <android.support.v7.widget.AppCompatTextView
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:layout_gravity="center"
            android:text="Toolbar" />

        </android.support.v7.widget.Toolbar>

        <FrameLayout
            android:id="@+id/container_view"
            android:layout_width="0dp"
            android:layout_height="0dp"
            app:layout_constraintBottom_toBottomOf="parent"
            app:layout_constraintEnd_toEndOf="parent"
            app:layout_constraintStart_toStartOf="parent"
            app:layout_constraintTop_toBottomOf="@+id/toolbar_view" />

    </android.support.constraint.ConstraintLayout>
 */
class ToolbarActivityInterceptor(
        callback: HasToolbarInterceptor
) : ActivityInterceptor<ToolbarInterceptor, HasToolbarInterceptor>(callback),
        ToolbarInterceptor {

    private var mToolbar: Toolbar? = null

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        super.onActivityCreated(activity, savedInstanceState)

        mToolbar = (activity as? AppCompatActivity)?.let {
            initToolbar(it)
        }
    }

    private fun initToolbar(activity: AppCompatActivity): Toolbar? = mCallback.onCreateToolbarView().also {
        activity.setSupportActionBar(it)
        activity.supportActionBar?.setDisplayShowTitleEnabled(false)
        it.setOnMenuItemClickListener { item -> activity.onOptionsItemSelected(item) }
        mCallback.onToolbarViewCreated(it)
    }
}
